File: src/main/java/dev/slne/discord/discord/interaction/command/commands/ticket/members/TicketMemberAddCommand.java
package dev.slne.discord.discord.interaction.command.commands.ticket.members;

import dev.slne.discord.annotation.DiscordCommandMeta;
import dev.slne.discord.discord.interaction.command.commands.TicketCommand;
import dev.slne.discord.exception.command.CommandException;
import dev.slne.discord.exception.ticket.member.TicketAddMemberException;
import dev.slne.discord.guild.permission.CommandPermission;
import dev.slne.discord.message.EmbedColors;
import dev.slne.discord.message.RawMessages;
import dev.slne.discord.spring.service.ticket.TicketService;
import dev.slne.discord.ticket.TicketChannelHelper;
import dev.slne.discord.ticket.member.TicketMember;
import dev.slne.discord.util.TimeUtils;
import java.util.List;
import java.util.Optional;
import javax.annotation.Nonnull;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.JDA;
import net.dv8tion.jda.api.entities.MessageEmbed;
import net.dv8tion.jda.api.entities.User;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.InteractionHook;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.api.interactions.commands.build.OptionData;
import org.jetbrains.annotations.NotNull;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.scheduling.annotation.Async;

/**
 * The type Ticket member add command.
 */
@DiscordCommandMeta(name = "add", description = "Füge einen Nutzer zu einem Ticket hinzu.", permission = CommandPermission.TICKET_ADD_USER)
public class TicketMemberAddCommand extends TicketCommand {

  private static final String USER_OPTION = "user";
  private final JDA jda;
  private final TicketChannelHelper ticketChannelHelper;

  @Autowired
  public TicketMemberAddCommand(JDA jda, TicketChannelHelper ticketChannelHelper,
      TicketService ticketService) {
    super(ticketService);
    this.jda = jda;
    this.ticketChannelHelper = ticketChannelHelper;
  }

  @Override
  public @Nonnull List<OptionData> getOptions() {
    return List.of(
        new OptionData(
            OptionType.USER,
            USER_OPTION,
            RawMessages.get("interaction.command.ticket.member.add.arg.member"),
            true,
            false
        )
    );
  }

  @Override
  public void internalExecute(SlashCommandInteractionEvent interaction, InteractionHook hook)
      throws CommandException {
    final User user = getUserOrThrow(interaction, USER_OPTION);

    hook.editOriginal(RawMessages.get("interaction.command.ticket.member.add.adding")).queue();

    if (user.equals(jda.getSelfUser())) {
      throw CommandException.ticketAddBot();
    }

    final Optional<TicketMember> ticketMember = getTicket().getActiveTicketMember(user);

    if (ticketMember.isPresent()) {
      throw CommandException.ticketAddMemberAlreadyInTicket();
    }

    try {
      addTicketMember(user, interaction.getUser(), hook);
    } catch (TicketAddMemberException e) {
      throw CommandException.ticketAddMember(e);
    }
  }

  @Async
  protected void addTicketMember(final User user, final User executor, InteractionHook hook)
      throws CommandException, TicketAddMemberException {
    final TicketMember newTicketMember = TicketMember.createFromTicket(getTicket(), user, executor);

    final TicketMember addedMember = getTicket().addTicketMember(newTicketMember).join();

    if (addedMember == null) {
      throw CommandException.ticketAddMember(new Throwable());
    }

    ticketChannelHelper.addTicketMember(getTicket(), addedMember).join();
    hook.editOriginal(RawMessages.get("interaction.command.ticket.member.add.added")).queue();
    getChannel().sendMessage(user.getAsMention())
        .setEmbeds(getAddedEmbed(executor))
        .queue();
  }

  /**
   * Returns the embed sent when a user is added to a ticket.
   *
   * @param adder the adder
   * @return The embed that is sent when a user is added to a ticket.
   */
  public MessageEmbed getAddedEmbed(@NotNull User adder) {
    return new EmbedBuilder()
        .setTitle(RawMessages.get("interaction.command.ticket.member.embed.title"))
        .setDescription(RawMessages.get("interaction.command.ticket.member.embed.description"))
        .setTimestamp(TimeUtils.berlinTimeProvider().getCurrentTime())
        .setColor(EmbedColors.ADD_TICKET_MEMBER)
        .setFooter(
            RawMessages.get("interaction.command.ticket.member.embed.footer", adder.getName()),
            adder.getAvatarUrl()
        )
        .build();
  }
}

File: src/main/kotlin/dev/slne/discord/discord/interaction/command/commands/ticket/members/TicketMemberAddCommand.kt
package dev.slne.discord.discord.interaction.command.commands.ticket.members

import dev.minn.jda.ktx.coroutines.await
import dev.slne.discord.annotation.DiscordCommandMeta
import dev.slne.discord.discord.interaction.command.checkMemberNotBot
import dev.slne.discord.discord.interaction.command.commands.TicketCommand
import dev.slne.discord.exception.command.CommandExceptions
import dev.slne.discord.guild.permission.CommandPermission
import dev.slne.discord.message.MessageManager
import dev.slne.discord.message.translatable
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.entities.channel.concrete.ThreadChannel
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.interactions.InteractionHook


private const val USER_OPTION = "user"

@DiscordCommandMeta(
    name = "add",
    description = "Füge einen Nutzer zu einem Ticket hinzu.",
    permission = CommandPermission.TICKET_ADD_USER
)
class TicketMemberAddCommand(private val messageManager: MessageManager) : TicketCommand() {
    override val options = listOf(
        option<Member>(
            USER_OPTION,
            translatable("interaction.command.ticket.member.add.arg.member"),
        )
    )

    override suspend fun internalExecute(
        interaction: SlashCommandInteractionEvent,
        hook: InteractionHook
    ) {
        val member = interaction.getOptionOrThrow<Member>(USER_OPTION)

        hook.editOriginal(translatable("interaction.command.ticket.member.add.checking-bot"))
            .await()
        member.checkMemberNotBot(CommandExceptions.TICKET_BOT_ADD)

        val thread = interaction.getThreadChannelOrThrow()

        hook.editOriginal(translatable("interaction.command.ticket.member.add.checking-already-in-ticket"))
            .await()
        val ticketMembers = thread.retrieveThreadMembers().await()

        if (ticketMembers.any { it.id == member.id }) {
            throw CommandExceptions.TICKET_MEMBER_ALREADY_IN_TICKET.create()
        }

        hook.editOriginal(translatable("interaction.command.ticket.member.add.adding")).await()
        addTicketMember(member, interaction.user, thread, hook)
    }

    private suspend fun addTicketMember(
        member: Member,
        executor: User,
        threadChannel: ThreadChannel,
        hook: InteractionHook
    ) {
        threadChannel.addThreadMember(member).await()

        hook.editOriginal(translatable("interaction.command.ticket.member.add.added")).await()
        hook.sendMessage(messageManager.buildMemberAddedMessage(member, executor)).await()
        hook.deleteOriginal().await()
    }
}
